### PR TITLE
Dns controller fixes

### PR DESCRIPTION
**Pull request by olemarkus**
My original intention was to try to correctly document how dns-controller worked. I had to test some code to check the actual behaviour, so may as well commit those too.

Ref #11055

### CODE REVIEW

File: dns-controller/pkg/watchers/testhelpers.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+type fakeScope struct {
+	readyCh chan struct{}
+	records map[string][]dns.Record
+}
+
+func (f *fakeScope) Replace(recordName string, records []dns.Record) {
+	f.records[recordName] = records
+}
+
+// MarkReady is called when a watcher has processed all inital resources
```
Comment (hakman) [inline dns-controller/pkg/watchers/testhelpers.go:30]:
```suggestion
// MarkReady is called when a watcher has processed all initial resources
```

File: dns-controller/README.md [hunk 2]
```diff
@@ -12,18 +14,41 @@ can end-users reach the API.  In future we might expose the API server
...
+#### Loadbalancer
+
+If _either_ of the two annotations are set on a `LoadBalancer` service, it will create a `CNAME` for the load balancer hostname _or_ it will create an A record if the load balancer has an IP.
+
+### Ingress 
+
+dns-controller can optionally watch `Ingress` resources. To enable this, you need to add the following to the cluster spec:
+```
+spec:
+  external-dns:
```
Comment (justinsb) [inline dns-controller/README.md:50]: I believe it is `externalDNS`

File: dns-controller/pkg/watchers/pod_test.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+	records, ok := scope.records["kube-system/somepod"]
+	if !ok {
+		t.Fatal("no records for pod generated")
+	}
+
+	if len(records) != 2 {
+		t.Fatalf("unexpected number of records")
+	}
+
```
Comment (justinsb) [inline dns-controller/pkg/watchers/pod_test.go:92]: Any reason not to reflect.DeepEqual or cmp.Diff on records here?   Feels like you did all the hard stuff :-)

**Comment by justinsb** [comment]
LGTM, thanks for this.  One small nit, and one suggestion!

File: dns-controller/pkg/watchers/pod_test.go (hunk 3 above)
Comment (olemarkus) [reply dns-controller/pkg/watchers/pod_test.go:92]: Hehe. I probably should have, but I got what I wanted from these tests around this point.

Comment (justinsb) [reply dns-controller/pkg/watchers/pod_test.go:92]: No worries - I'll merge and send a PR with these two suggestions then!

**Comment by justinsb** [comment]
Thanks @olemarkus 

/approve
/lgtm

**Comment by justinsb** [comment]
/retest

apiserver flake

**Pull request merged**